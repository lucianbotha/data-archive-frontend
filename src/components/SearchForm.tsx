import * as React from "react";
import {
  CalibrationType,
  IGeneral,
  ISearchFormState,
  ITarget,
  ITelescope
} from "../utils/ObservationQueryParameters";
import {
  ButtonGrid,
  DataGrid,
  ParentGrid,
  ProposalGrid,
  Spinner,
  TargetGrid,
  TelescopeGrid
} from "./basicComponents/Grids";
import DataForm from "./searchFormComponents/DataForm";
import ProposalForm, {
  validatedProposal
} from "./searchFormComponents/ProposalForm";
import TargetForm, { validatedTarget } from "./searchFormComponents/TargetForm";
import TelescopeForm, {
  validatedTelescope
} from "./searchFormComponents/TelescopeForm";

/**
 * A form for defining search parameters for an observation search, and for
 * initiating the search.
 */
class SearchForm extends React.Component<{}, ISearchFormState> {
  public state: ISearchFormState = {
    general: { calibrations: new Set<CalibrationType>(), errors: {} },
    loading: false,
    target: {
      errors: {},
      resolver: "Simbad",
      searchConeRadiusUnits: "arcseconds"
    }
  };

  /**
   * Handle changes of telescope-related parameters.
   */
  public telescopeChange = (value: ITelescope) => {
    const newState = {
      ...this.state,
      telescope: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Handle changes of target-related parameters.
   */
  public targetChange = (value: ITarget) => {
    const newState = {
      ...this.state,
      target: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Handle changes of general parameters.
   */
  public generalChange = (value: IGeneral) => {
    const newState = {
      ...this.state,
      general: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Perform an observation with the currently selected search parameters.
   */
  public searchArchive = async () => {
    this.setState(() => ({
      ...this.state,
      loading: true
    }));

    // Add errors to the search parameter details
    const target = await validatedTarget(this.state.target);
    const general = validatedProposal(this.state.general);
    const telescope = validatedTelescope(this.state.telescope);

    this.setState(() => ({
      ...this.state,
      general,
      loading: false,
      target,
      telescope
    }));
  };

  public render() {
    const { target, general, telescope, loading } = this.state;

    return (
      <>
        {loading && (
          <div className="dimScreen">
            <Spinner />
          </div>
        )}
        <ParentGrid>
          <TargetGrid>
            <TargetForm target={target} onChange={this.targetChange} />
          </TargetGrid>
          <ProposalGrid>
            <ProposalForm proposal={general} onChange={this.generalChange} />
          </ProposalGrid>
          <TelescopeGrid>
            <TelescopeForm
              telescope={telescope}
              onChange={this.telescopeChange}
            />
          </TelescopeGrid>
          <DataGrid>
            <DataForm general={general} onChange={this.generalChange} />
          </DataGrid>
          <ButtonGrid>
            <input
              className="button is-primary"
              type="button"
              value="Search"
              onClick={this.searchArchive}
            />
          </ButtonGrid>
        </ParentGrid>
      </>
    );
  }
}

export default SearchForm;
