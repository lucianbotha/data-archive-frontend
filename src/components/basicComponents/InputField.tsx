import * as React from "react";
import styled from "styled-components";

interface IInputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  error?: string | undefined;
}

/**
 * A paragraph with an error message.
 */
const Error = styled.p.attrs({
  className: "has-text-danger"
})``;

/**
 * An input field with an error message underneath it.
 *
 * The field accepts all attributes you can pass to an HTML input element. In
 * addition, you may pass an error argument with an error message to display.
 */
const InputField = (props: IInputProps) => {
  const { className, error, ...rest } = props;
  return (
    <div className="control">
      <input
        type={"text"}
        className={`input ${className} ${error && error !== "" && "is-danger"}`}
        {...rest}
      />
      {error && <Error>{error}</Error>}
    </div>
  );
};
export default InputField;
