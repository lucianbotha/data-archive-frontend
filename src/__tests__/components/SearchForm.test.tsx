import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import SearchForm from "../../components/SearchForm";
import InputField from "../../components/searchFormComponents/TargetForm";

describe("Search Form", () => {
  const wrapper = mount(<SearchForm />);
  it("should render", () => {
    expect(wrapper).toBeDefined();
  });

  it("Should always renders a div of class name grid-container", () => {
    const mainDiv = wrapper.find("div");
    expect(mainDiv.length).toBeGreaterThan(0);
    const mainContainer = wrapper.find("div.grid-container");
    expect(mainContainer.length).toEqual(1);
  });

  it("should contain five grid-item and one button", () => {
    const items = wrapper.find("div.grid-item");
    expect(items.length).toEqual(5);
    const button = wrapper.find("input.is-primary");
    expect(button.length).toEqual(1);
  });

  it("should contain target, proposal, telescope and data form", () => {
    const target = wrapper.find("div.target-form");
    expect(target.length).toEqual(1);
    const proposal = wrapper.find("div.proposal-form");
    expect(proposal.length).toEqual(1);
    const telescope = wrapper.find("div.telescope-form");
    expect(telescope.length).toEqual(1);
    const data = wrapper.find("div.data-form");
    expect(data.length).toEqual(1);
  });

  it("Should update state when typing", () => {
    let value: any;

    const targetInput = wrapper.find('.input[data-test="target-name-input"]');
    const targetName = targetInput.find("input");
    console.log(targetInput);
    console.log(targetName);

    value = "apple";
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.state() as any).target.name).toEqual("apple");

    value = "";
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.state() as any).target.name).toEqual("");

    value = undefined;
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.state() as any).target.name).toEqual("");

    value = null;
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.state() as any).target.name).toEqual("");

    value = "\\ \t";
    targetName.simulate("change", { target: { value, name: "name" } });
    expect((wrapper.state() as any).target.name).toEqual("\\ \t");
  });

  it("Should update state when change a select", () => {
    let value: any;
    const resolverSelect = wrapper.find('select[data-test="resolver-select"]');
    const resolver = resolverSelect.find("select");
    value = "NED";
    const event = { target: { value, name: "resolver" } };
    resolver.simulate("change", event);
    expect((wrapper.state() as any).target.resolver).toEqual("NED");
  });

  it("render correctly", () => {
    expect(toJson(shallow(<SearchForm />))).toMatchSnapshot();
  });
});
