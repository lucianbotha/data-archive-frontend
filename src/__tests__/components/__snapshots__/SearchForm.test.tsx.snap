// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form render correctly 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <TargetForm
        onChange={[Function]}
        target={
          Object {
            "errors": Object {},
          }
        }
      />
    </styled.div>
    <styled.div>
      <ProposalForm
        onChange={[Function]}
        proposal={
          Object {
            "errors": Object {},
          }
        }
      />
    </styled.div>
    <styled.div>
      <TelescopeForm
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <DataForm
        data={
          Object {
            "errors": Object {},
          }
        }
        onChange={[Function]}
      />
    </styled.div>
    <styled.div>
      <input
        className="button is-primary"
        onClick={[Function]}
        type="button"
        value="Search"
      />
    </styled.div>
  </styled.div>
</Fragment>
`;
